### PR TITLE
[IMPROVED] Stalled (queue) subscriptions

### PR DESCRIPTION
**Pull request by kozlovic**
- When new message arrives in channel and sent to subscription,
  a lookup was done even if the subscription was stalled. This
  is no longer the case.
- A queue subscription is stalled only when all its members are
  stalled. Again, no lookup necessary if the group is stalled.
- Removal of "stalled" debug statement that would be printed
  for every message after a subscription has reached its
  MaxInflight.

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -415,7 +415,7 @@ type queueState struct {
 	lastSent  uint64
 	subs      []*subState
 	shadow    *subState // For durable case, when last member leaves and group is not closed.
-	stalled   bool
+	stalled   int       // number of stalled members
```
Comment (petemiron) [inline server/server.go]: recommend making this variable name more descriptive `stalledMemberCnt` or `stalledSubCnt`. parsing some of the latter logic of subscriber stalled flag vs. this stalled var becomes hard to follow.

File: server/server.go [hunk 2]
```diff
@@ -3430,10 +3427,14 @@ func (s *StanServer) processAck(c *channel, sub *subState, sequence uint64) {
 		return
 	}
 
-	delete(sub.acksPending, sequence)
-	stalled := sub.stalled
-	if int32(len(sub.acksPending)) < sub.MaxInFlight {
-		sub.stalled = false
+	// Do this here only if this is not a queue member. If it is, it has
```
Comment (petemiron) [inline server/server.go]: consider setting the `qs` variable before this `if`, then converting this to an `if/else` instead of 2 separate `if`s.

File: server/server.go [hunk 3]
```diff
@@ -3443,8 +3444,21 @@ func (s *StanServer) processAck(c *channel, sub *subState, sequence uint64) {
 
 	if qs != nil {
 		qs.Lock()
-		stalled = qs.stalled
-		qs.stalled = false
+		stalled = qs.stalled == len(qs.subs)
```
Comment (petemiron) [inline server/server.go]: specific example of hard-to-parse logic.

File: server/server.go (hunk 2 above)
Comment (kozlovic) [inline server/server.go]: Note that the sub's lock need to be released before grabbing the queue state's lock. Still, I will update based on your comment and you will decide if it is better or not.

File: server/server.go (hunk 1 above)
Comment (kozlovic) [inline server/server.go]: I did that originally, so I agree. I'll go for `stalledSubCount`.

File: server/server.go (hunk 3 above)
Comment (kozlovic) [inline server/server.go]: agreed
